src/Datadog.Trace/Propagation/B3/B3SpanContextPropagator.cs

```diff
@@ -0,0 +1,209 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using Datadog.Trace.Headers;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Propagation.B3
+{
+    /// <summary>
+    /// Class that hanbles B3 style context propagation.
+    /// </summary>
+    internal class B3SpanContextPropagator : Propagator
+    {
+        private const NumberStyles NumberStyle = NumberStyles.HexNumber;
+        private static readonly string UserKeep = ((int)SamplingPriority.UserKeep).ToString(CultureInfo.InvariantCulture);
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<B3SpanContextPropagator>();
+
+        private B3SpanContextPropagator()
+        {
+        }
+
+        /// <summary>
+        /// Gets the singleton instance of the propagator.
+        /// </summary>
+        public static B3SpanContextPropagator Instance { get; } = new B3SpanContextPropagator();
```

Comment:
Maybe we can get rid of this singleton?

Reply:
Yes, the propagator is now coupled with Tracer, as Tracer is coupled with settings and settings define the propagator. So for me seems logical refactor.